pkg/plugins/datasource.go

```diff
@@ -69,26 +87,48 @@ func DataSources(l logger.Interface, s schema.Stage, fs vfs.FS, console Console)
 
 	var p prv.Provider
 	var userdata []byte
-	var err error
-	found := false
+
+	userdataDone := make(chan []byte, len(uniqueProviders))
+	var wg sync.WaitGroup
+	if len(AvailableProviders) > 0 {
+		l.Debugf("Full provider list: %s", AvailableProviders)
+	}
+
 	for _, p = range AvailableProviders {
-		if p.Probe() {
-			userdata, err = p.Extract()
-			if err != nil {
-				l.Warnf("Failed extracting data from %s provider: %s", p.String(), err.Error())
+		l.Debugf("Starting provider %s", p.String())
+		prov := p
+		wg.Add(1)
+		go func(l logger.Interface, p prv.Provider) {
```

Comment:
maybe we should run in sequence only cdrom - I'm thinking, you could have cdrom and datasource together, and in with parallel execution is not predictable what happens (who's going to win?)

Reply:
is this a feasible scenario? having 2 datasources one being cdrom and one being a metadata?

Reply:
updated with this, we can drop it afterwards if we find this is not useful. I find it quite strange to have 2 sources of metadata but indeed it could happen in something like openstack.

Reply:
I think could likely happen that, for instance you use a cidata on openstack, let's say, and openstack serving its own datasources. In that case what would happen is - first cdrom, and then metadata. Maybe we should add a guard in that case to not override the userdata with empty files

Reply:
in any case, this will now just only run the providers if the userdata is not emtpy after going trougth cdrom, so it should be safe?